modules/e2ee/ManagedKeyHandler.js

```diff
@@ -0,0 +1,183 @@
+ */
+export class ManagedKeyHandler extends KeyHandler {
+    /**
+     * Build a new AutomaticKeyHandler instance, which will be used in a given conference.
+     */
+    constructor(conference) {
+        super(conference);
+
+        this._key = undefined;
+        this._conferenceJoined = false;
+
+        this._olmAdapter = new OlmAdapter(conference);
+
+        this._rotateKey = debounce(this._rotateKeyImpl, DEBOUNCE_PERIOD);
+        this._ratchetKey = debounce(this._ratchetKeyImpl, DEBOUNCE_PERIOD);
+
+        // Olm signalling events.
+        this._olmAdapter && this._olmAdapter.on(
+            OlmAdapter.events.PARTICIPANT_KEY_UPDATED,
+            this._onParticipantKeyUpdated.bind(this));
+
+        this.conference.on(
+            JitsiConferenceEvents.PARTICIPANT_PROPERTY_CHANGED,
+            this._onParticipantPropertyChanged.bind(this));
+        this.conference.on(
+            JitsiConferenceEvents.USER_JOINED,
+            this._onParticipantJoined.bind(this));
+        this.conference.on(
+            JitsiConferenceEvents.USER_LEFT,
+            this._onParticipantLeft.bind(this));
+        this.conference.on(
+                JitsiConferenceEvents.CONFERENCE_JOINED,
+                () => {
+                    this._conferenceJoined = true;
+                });
+    }
+
+    /**
+     * When E2EE is enabled it initializes sessions and sets the key.
+     * Cleans up the sessions when disabled.
+     *
+     * @param {boolean} enabled - whether E2EE should be enabled or not.
+     * @returns {void}
+     */
+    async _setEnabled(enabled) {
+        if (enabled) {
+            await this._olmAdapter.initSessions();
+        } else {
+            this._olmAdapter.clearAllParticipantsSessions();
+        }
+
+        // Generate a random key in case we are enabling.
+        this._key = enabled ? this._generateKey() : false;
+
+        // Send it to others using the E2EE olm channel.
+        const index = await this._olmAdapter.updateKey(this._key);
+
+        // Set our key so we begin encrypting.
+        this.e2eeCtx.setKey(this.conference.myUserId(), this._key, index);
+    }
+
+    /**
+     * Handles an update in a participant's presence property.
+     *
+     * @param {JitsiParticipant} participant - The participant.
+     * @param {string} name - The name of the property that changed.
+     * @param {*} oldValue - The property's previous value.
+     * @param {*} newValue - The property's new value.
+     * @private
+     */
+    async _onParticipantPropertyChanged(participant, name, oldValue, newValue) {
+        switch (name) {
+        case 'e2ee.idKey':
+            logger.debug(`Participant ${participant.getId()} updated their id key: ${newValue}`);
+            break;
+        case 'e2ee.enabled':
+            if (!newValue && this.enabled) {
+                this._olmAdapter && this._olmAdapter.clearParticipantSession(participant);
+
+                this._rotateKey();
```

Comment:
Not related to this change, but can you please remind me: why did we do this? Looks like we are rotating the key when a participant disables e2ee while we have it enabled? Why is that?